client/allocrunner/tasklifecycle/gate.go

```diff
@@ -0,0 +1,80 @@
+package tasklifecycle
+
+const (
+	gateClosed = false
+	gateOpened = true
+)
+
+// Gate is used by the Coordinator to block or allow tasks from running.
+//
+// It provides a channel that taskRunners listens on to determine when they are
+// allowed to run. The Gate has an infinite loop that is either feeding this
+// channel (therefore allowing listeners to proceed) or not doing anything
+// (causing listeners to block an wait).
+//
+// The Coordinator uses the Gate Open() and Close() methods to control this
+// producer loop.
+type Gate struct {
+	sendCh     chan struct{}
+	updateCh   chan bool
+	shutdownCh <-chan struct{}
+}
+
+// NewGate returns a new Gate that is initially closed. The Gate should not be
+// used after the shutdownCh is closed.
+func NewGate(shutdownCh <-chan struct{}) *Gate {
+	g := &Gate{
+		sendCh:     make(chan struct{}),
+		updateCh:   make(chan bool),
+		shutdownCh: shutdownCh,
+	}
+	go g.run(gateClosed)
+
+	return g
+}
+
+// WaitCh returns a channel that the listener must block on before starting its
+// task.
+func (g *Gate) WaitCh() <-chan struct{} {
+	return g.sendCh
+}
+
+// Open is used to allow listeners to proceed.
+func (g *Gate) Open() {
+	select {
+	case <-g.shutdownCh:
+	case g.updateCh <- gateOpened:
+	}
+}
+
+// Close is used to block listeners from proceeding.
+func (g *Gate) Close() {
+	select {
+	case <-g.shutdownCh:
+	case g.updateCh <- gateClosed:
+	}
+}
```

Comment:
Perhaps include in the comment what the shutdown behavior is as if the callers cares about whether they were unblocked due to Open/Close vs Shutdown, they'll need to check for the shutdown condition themselves.

A quick peek didn't reveal any callers that would care about Open/Close vs Shutdown, but I think a comment would still be nice in case this code is used more broadly in the future.

Reply:
Not sure if this is what you were thinking about, but I added these lines:

```go
// if the gate shutdownch channel is closed, this method is a no-op so callers
// should check its state.
```